src/disk.c

```diff
@@ -660,36 +686,53 @@ static int disk_read(void) {
     if (ignorelist_match(ignorelist, disk_name) != 0)
       continue;
 
+    metric_t m = {0};
+    metric_label_set(&m, "device", disk_name);
+
     if ((snap_iter->bytes[DEVSTAT_READ] != 0) ||
         (snap_iter->bytes[DEVSTAT_WRITE] != 0)) {
-      disk_submit(disk_name, "disk_octets",
-                  (derive_t)snap_iter->bytes[DEVSTAT_READ],
-                  (derive_t)snap_iter->bytes[DEVSTAT_WRITE]);
+      m.value.counter = (counter_t)snap_iter->bytes[DEVSTAT_READ];
+      metric_family_metric_append(&fam_disk_read_bytes, m);
+
+      m.value.counter = (counter_t)snap_iter->bytes[DEVSTAT_WRITE];
+      metric_family_metric_append(&fam_disk_write_bytes, m);
     }
 
     if ((snap_iter->operations[DEVSTAT_READ] != 0) ||
         (snap_iter->operations[DEVSTAT_WRITE] != 0)) {
-      disk_submit(disk_name, "disk_ops",
-                  (derive_t)snap_iter->operations[DEVSTAT_READ],
-                  (derive_t)snap_iter->operations[DEVSTAT_WRITE]);
+      m.value.counter = (counter_t)snap_iter->operations[DEVSTAT_READ];
+      metric_family_metric_append(&fam_disk_read_ops, m);
+
+      m.value.counter = (counter_t)snap_iter->operations[DEVSTAT_WRITE];
+      metric_family_metric_append(&fam_disk_write_ops, m);
     }
 
     read_time = devstat_compute_etime(&snap_iter->duration[DEVSTAT_READ], NULL);
     write_time =
         devstat_compute_etime(&snap_iter->duration[DEVSTAT_WRITE], NULL);
     if ((read_time != 0) || (write_time != 0)) {
-      disk_submit(disk_name, "disk_time", (derive_t)(read_time * 1000),
-                  (derive_t)(write_time * 1000));
+      m.value.counter = (counter_t)(read_time * 1000);
+      metric_family_metric_append(&fam_disk_read_time, m);
+
+      m.value.counter = (counter_t)(write_time * 1000);
+      metric_family_metric_append(&fam_disk_write_time, m);
     }
     if (devstat_compute_statistics(snap_iter, NULL, 1.0, DSM_TOTAL_BUSY_TIME,
                                    &busy_time, DSM_TOTAL_DURATION,
                                    &total_duration, DSM_QUEUE_LENGTH,
                                    &queue_length, DSM_NONE) != 0) {
       WARNING("%s", devstat_errbuf);
     } else {
-      submit_io_time(disk_name, busy_time, total_duration);
-      submit_in_progress(disk_name, (gauge_t)queue_length);
+      m.value.counter = busy_time;
```

Comment:
Cast from `double` to `counter_t`, the same as for `write_time` and `read_time`.

Reply:
Casting added.